src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipeline.cs

```diff
@@ -13,18 +13,20 @@
 using Validation;
 
 using Microsoft.Diagnostics.EventFlow.Configuration;
+using System.Diagnostics;
 
 namespace Microsoft.Diagnostics.EventFlow
 {
     public class DiagnosticPipeline : IDisposable
     {
-        private List<IDisposable> pipelineDisposables;
+        private List<IDisposable> pipelineLinkDisposables;
         private List<IDisposable> inputSubscriptions;
         private bool disposed;
         private bool disposeDependencies;
         private CancellationTokenSource cancellationTokenSource;
         private IDataflowBlock pipelineHead;
         private DiagnosticPipelineConfiguration pipelineConfiguration;
+        private volatile int eventsInProgress;
```

Comment:
Ideally, yes.

If we do not use completion propagation, complete from head to tail, and wait for completion on each block, that would give us what we want, i.e. drain the pipeline ensuring that each event is processed fully.

But that would be a bit tricky code to walk the pipeline and also account for how much time is left before we hit the timeout.

I think counting the events in the pipeline is simpler.